src/Microsoft.VisualStudio.Services.Agent/AgentServer.cs

```diff
@@ -41,44 +54,120 @@ public interface IAgentServer : IAgentService
 
     public sealed class AgentServer : AgentService, IAgentServer
     {
-        private bool _hasConnection;
-        private VssConnection _connection;
-        private TaskAgentHttpClient _taskAgentClient;
-
-        public async Task ConnectAsync(VssConnection agentConnection)
+        private bool _hasGenericConnection;
+        private bool _hasMessageConnection;
+        private bool _hasRequestConnection;
+        private VssConnection _genericConnection;
+        private VssConnection _messageConnection;
+        private VssConnection _requestConnection;
+        private TaskAgentHttpClient _genericTaskAgentClient;
+        private TaskAgentHttpClient _messageTaskAgentClient;
+        private TaskAgentHttpClient _requestTaskAgentClient;
+
+        public async Task ConnectAsync(Uri serverUrl, VssCredentials credentials)
         {
             if (HostContext.RunMode == RunMode.Local)
             {
                 return;
             }
 
-            _connection = agentConnection;
+            _genericConnection = await EstablishVssConnection(serverUrl, credentials, TimeSpan.FromSeconds(100));
+            _messageConnection = await EstablishVssConnection(serverUrl, credentials, TimeSpan.FromSeconds(60));
+            _requestConnection = await EstablishVssConnection(serverUrl, credentials, TimeSpan.FromSeconds(60));
+
+            _genericTaskAgentClient = _genericConnection.GetClient<TaskAgentHttpClient>();
+            _messageTaskAgentClient = _messageConnection.GetClient<TaskAgentHttpClient>();
+            _requestTaskAgentClient = _requestConnection.GetClient<TaskAgentHttpClient>();
+
+            _hasGenericConnection = true;
+            _hasMessageConnection = true;
+            _hasRequestConnection = true;
+        }
+
+        public async Task RefreshConnectionAsync(AgentConnectionType connectionType, TimeSpan timeout)
```

Comment:
Is this never called for `Generic` because of the parallel nature of the generic consumers?

And message queue and job request are always called in serial (do the queues enforce this?)? Are there any conditions where they are not serial? Is there a race condition if not where hasConnection can be false due to being reset or failure during reset?

It makes me wonder whether a different approach would make more sense. Something more like a factory, that establishes a connection as needed and caches successful connections.

I think I could go either way, but fuzzy on the behavior of consumers. It feels like this implementation is making a lot of assumptions about it's consumers (and may be justified). Let's discuss in the morning and we can talk through this implementation and the merits of a potential alternative I have in mind.